core/trino-main/src/main/java/io/trino/operator/hash/fastbb/VarHandleFastByteBuffer.java

```diff
@@ -0,0 +1,107 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.operator.hash.fastbb;
+
+import io.airlift.slice.Slice;
+import io.airlift.slice.Slices;
+
+import java.lang.invoke.MethodHandles;
+import java.lang.invoke.VarHandle;
+import java.nio.ByteOrder;
+
+public class VarHandleFastByteBuffer
+        implements FastByteBuffer
+{
+    private static final VarHandle LONG_ARRAY_HANDLE = MethodHandles.byteArrayViewVarHandle(long[].class, ByteOrder.nativeOrder());
+    private static final VarHandle INT_ARRAY_HANDLE = MethodHandles.byteArrayViewVarHandle(int[].class, ByteOrder.nativeOrder());
+    private static final VarHandle SHORT_ARRAY_HANDLE = MethodHandles.byteArrayViewVarHandle(short[].class, ByteOrder.nativeOrder());
+
+    private final byte[] array;
```

Comment:
This implementation limits possible number of groups. If the aggregation is made using several long fields it can easily get to 100B / row making this array unusable above 20M groups (2^31B / 100B).

Reply:
that's a valid point.
This is not as bad because for partial aggregation we will never hit 20M because of the memory limit and the final step is partitioned so the total number of groups is ~ groups per single hash table X number of nodes x number of cores (e.g. ~3B for 6 nodes with 32 cores).
Also, 100 bytes is for 8 bigints so not the common case.
That said it needs to be handled. One way is to switch to OffHeapFastByteBuffer once we hit this limit (we would have to manage off heap then). Another is to switch to the old columnar implementation.